src/portable/portable.c

```diff
@@ -1559,40 +1559,54 @@ int portable_detach(
         }
 
         FOREACH_DIRENT(de, d, return log_debug_errno(errno, "Failed to enumerate '%s' directory: %m", where)) {
-                _cleanup_free_ char *marker = NULL;
+                _cleanup_free_ char *marker = NULL, *unit_name = NULL;
                 UnitFileState state;
+                const char *dot;
+
+                dot = endswith(de->d_name, ".d");
+                if (dot)
+                        unit_name = strndup(de->d_name, dot - de->d_name);
+                else
+                        unit_name = strdup(de->d_name);
+                if (!unit_name)
+                        return -ENOMEM;
 
-                if (!unit_name_is_valid(de->d_name, UNIT_NAME_ANY))
+                if (!unit_name_is_valid(unit_name, UNIT_NAME_ANY))
                         continue;
 
                 /* Filter out duplicates */
-                if (set_contains(unit_files, de->d_name))
+                if (set_contains(unit_files, unit_name))
                         continue;
 
-                if (!IN_SET(de->d_type, DT_LNK, DT_REG))
+                if (dot ? !IN_SET(de->d_type, DT_LNK, DT_DIR) : !IN_SET(de->d_type, DT_LNK, DT_REG))
                         continue;
 
-                r = test_chroot_dropin(d, where, de->d_name, name_or_path, extension_image_paths, &marker);
+                if (dot && laccess(unit_name, F_OK) >= 0)
```

Comment:
This is not necessary, no? This block only deals with the unit name, so whether it's sourced from the drop-in dir or the unit it doesn't change anything, it will be stored in the set anyway